group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -1805,6 +2019,290 @@ private List<ShareGroupHeartbeatResponseData.TopicPartitions> fromShareGroupAssi
+     * 4) Update the target assignment for the streams group if the group epoch
+     *    is larger than the current target assignment epoch.
+     * 5) Reconcile the member's assignment with the target assignment.
+     *
+     * @param groupId             The group ID from the request.
+     * @param memberId            The member ID from the request.
+     * @param memberEpoch         The member epoch from the request.
+     * @param instanceId          The instance ID from the request or null.
+     * @param rackId              The rack ID from the request or null.
+     * @param rebalanceTimeoutMs  The rebalance timeout from the request or -1.
+     * @param clientId            The client ID.
+     * @param clientHost          The client host.
+     * @param topology            The topology from the request or null.
+     * @param ownedActiveTasks    The list of owned active tasks from the request or null.
+     * @param ownedStandbyTasks   The list of owned standby tasks from the request or null.
+     * @param ownedWarmupTasks    The list of owned warmup tasks from the request or null.
+     * @param userEndpoint        User-defined endpoint for Interactive Queries, or null.
+     * @param clientTags          Used for rack-aware assignment algorithm, or null.
+     * @param taskEndOffsets      Cumulative changelog offsets for tasks, or null.
+     * @param taskOffsets         Cumulative changelog end-offsets for tasks, or null.
+     * @param shutdownApplication Whether all Streams clients in the group should shut down.
+     * @return A result containing the StreamsGroupHeartbeat response and a list of records to update the state machine.
+     */
+    private CoordinatorResult<StreamsGroupHeartbeatResult, CoordinatorRecord> streamsGroupHeartbeat(
+        String groupId,
+        String memberId,
+        int memberEpoch,
+        String instanceId,
+        String rackId,
+        int rebalanceTimeoutMs,
+        String clientId,
+        String clientHost,
+        StreamsGroupHeartbeatRequestData.Topology topology,
+        List<TaskIds> ownedActiveTasks,
+        List<TaskIds> ownedStandbyTasks,
+        List<TaskIds> ownedWarmupTasks,
+        String processId,
+        Endpoint userEndpoint,
+        List<KeyValue> clientTags,
+        List<TaskOffset> taskOffsets,
+        List<TaskOffset> taskEndOffsets,
+        boolean shutdownApplication
+    ) throws ApiException {
+        final long currentTimeMs = time.milliseconds();
+        final List<CoordinatorRecord> records = new ArrayList<>();
+        final List<StreamsGroupHeartbeatResponseData.Status> returnedStatus = new ArrayList<>();
+
+        // Get or create the streams group.
+        boolean isJoining = memberEpoch == 0;
+        final StreamsGroup group = isJoining ? getOrCreateStreamsGroup(groupId) : getStreamsGroupOrThrow(groupId);
+        throwIfStreamsGroupIsFull(group, memberId);
```

Comment:
Overall, I'm fine with this but why not have a check for a full group in `getStreamsGroupOrThrow`?

Reply:
I think there would be a risk that, in the future, somebody uses `getStreamsGroupOrThrow` in a place where `GroupMaxSizeReachedException` is not expected. It would be a bit unexpected that `getStreamsGroupOrThrow` also does validation that is only needed if we plan to add a member. I think it's better to keep those concerns separate.

Reply:
Should we call this only if `isJoining` is true? And if yes, could we simplify the logic of `throwIfStreamsGroupIsFull` and only check its size? -- If `isJoining` is false, I would assume that `memberId` is never empty and that the `memberId` is never new (ie, always contained in `group`?